sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/BootstrapperImpl.java

```diff
@@ -32,45 +35,47 @@ public BootstrapperImpl(PartitionSynchronizer synchronizer, LeaseStore leaseStor
         this.leaseStore = leaseStore;
         this.lockTime = lockTime;
         this.sleepTime = sleepTime;
+
+        this.isInitialized = false;
     }
 
     @Override
     public Mono<Void> initialize() {
-        BootstrapperImpl self = this;
-
-        return Mono.fromRunnable( () -> {
-            while (true) {
-                boolean initialized = self.leaseStore.isInitialized().block();
+        this.isInitialized = false;
 
-                if (initialized) break;
+        return Mono.just(this)
+            .flatMap( value -> this.leaseStore.isInitialized())
+            .flatMap(initialized -> {
+                this.isInitialized = initialized;
 
-                boolean isLockAcquired = self.leaseStore.acquireInitializationLock(self.lockTime).block();
+                if (initialized) {
+                    return Mono.empty();
+                } else {
+                    return this.leaseStore.acquireInitializationLock(this.lockTime)
+                        .flatMap(lockAcquired -> {
+                            this.isLockAcquired = lockAcquired;
```

Comment:
the same question on thread-safety here for the field (inInitialized and isLockAccquried)

when you use flatMap, etc, that could technically change the thread so the thread accessing isLockAccquired may not be the same as the next thread in the FlatMap reading isLockAccquired.

I am expert on CFP logic flow, but it seems to me a class field is modified on different threads here.

is this ok?

Reply:
Yes, there's only one BootstrapperImpl instance per each ChangeFeedProcessorBuilderImpl. In this particular case the real lock is acquired via a specific document that is written in the lease collection which ensure only one Bootstrapper can create the lease documents.